noir-projects/noir-contracts/contracts/card_game_contract/src/cards.nr

```diff
@@ -120,9 +120,9 @@ impl Deck<&mut PrivateContext> {
         inserted_cards
     }
 
-    pub fn get_cards<N>(&mut self, cards: [Card; N]) -> [CardNote; N] {
+    pub fn remove_cards<N>(&mut self, cards: [Card; N]) {
```

Comment:
We never used the get_cards function out of the original removed cards so I just merged these 2 so that I could use pop_notes.

Looking at the function more I think there is a big inefficiency here now. The `filter_card` filter is applied in a constrained context (see constrain_get_notes_internal func) and therefore the notes obtained below are guaranteed to be properly filtered. Properly filtered here means that all the card notes returned correspond to some of the card from the cards arg. This however doesn't mean that all the cards from the cards arg were found. But it seems to me that it would be enough to just check that the num notes returned is equal to N.

Since this contract is quite old I think it's possible that it was implemented in a time were filters were not constrained or smt. like that.

Am I missing something here?

Bothering @sirasistant as you wrote the contract.

Reply:
Yes, I had noticed the same thing when I tried to do this some time ago. Keep in mind that we only started constraining the filter in #6703 - it was unconstrained before that.

Ultimately though I would not worry too much about this sample contract.

Reply:
Sure, but I feel like the contracts are used as an example for community so would say makes sense keeping them somewhat updated.

Reply:
Agreed. I went over this and I think you're right - just checking that the number of notes equals `N` should be enough here, since you only get the requested cards, and you only get each card once.

The only thing that gives me pause here is that we'll be applying the filter to the entire set of 32 possible notes, which seems wasteful if we later set a lower limit. Perhaps we can later revisit `get_notes` to get it's max number of notes to be generic instead of the actual maximum.

Reply:
>The only thing that gives me pause here is that we'll be applying the filter to the entire set of 32 possible notes, which seems wasteful if we later set a lower limit. Perhaps we can later revisit get_notes to get it's max number of notes to be generic instead of the actual maximum.

Nice observation. Shall we create an issue for this?